### PR TITLE
💄 Add Dark Mode

### LINKED ISSUE #406: 🌙 DarkMode
I want to spare my eyes as I pursue my passion for travel in the dark nights, swinging from train to train. A dark mode would help there. All black. `#000`.

### PR DESCRIPTION
**Pull request by thilobillerbeck**
closes #406 

At GPN I started to get the idea of implementing a dark mode since this is a common feature these days in web apps and will improve the usability of the site for people who prefer darker color schemes. Thus I started implementing a dark color scheme in combination with a toggle mechanism to manually switch between them. By default, the system preference will be picked.

The color scheme mainly focusses on a good readability with good but low contrasts to be easy on the eyes, espeacially in darker environments where bright colors can produce a lot of eyestrain. Meanwhile I tried to maintain the essence of the traewelling design to ensure people have the same experience as in light mode.

This is nearly finished but I still have to clean up some quirks, the code and maybe improve the switching mechanism. Meanwhile I want your help to improve everything in the last steps. This PR can be seen as a place to test and bring up some feedback to ensure a good experience for as many people as possible. 

Some questions to maybe help you giving feedback:

- How do you like the color scheme? What would you see as an improvement?
- What do you think about the switching mechanism? Is it okay like that or should it go into a drop down menu like the language selection?
- How do you feel about eyestrain while reading through the site? Especially in more crowded contexts like the feed.

### CODE REVIEW

**Comment by HerrLevin** [comment]
One thing about the switching mechanism: I really like what getbootstrap.com is doing in their dropdown. They also include the possibility of "auto" so that the system settings can be (automatically) used. I really love that feature and I think many people will prefer a webpage/app to automatically respect their system preference.
image

**Comment by MrKrisKrisu** [comment]
First of all: Thank you for your work! I love to see, that Trwl is getting an dark mode because of you. <3 

> How do you like the color scheme? What would you see as an improvement?

In general, I am very satisfied with the DarkMode. I find the readability of the red (trwl red) text on the dark background a bit difficult, but I would put that aside at this point unless you have an idea. Currently I have no idea how to make the trwl red fit better there. I also found some small things that are not perfect to read, like some buttons.

image

Also I am not fully happy with the color difference between the top nav bar, the jumbotron (e.g. on the profile page or the help page) and the dark background. But that are also some things to optimize maybe later step by step?

image

Some messages (e.g. the "Checked in" message) are also still a bit light.

image


> What do you think about the switching mechanism? Is it okay like that or should it go into a drop down menu like the language selection?

I would prefer a dropdown like @HerrLevin suggested. Maybe this would fit into the users dropdown in the top navbar?

> How do you feel about eyestrain while reading through the site? Especially in more crowded contexts like the feed.

Readability is good! As mentioned before, the only thing that bothers me a bit is the red text on a dark background.




I also noticed that when reloading a page or clicking on it, the dark mode is active by default (css class "dark" is set by default in the HTML). But if you use the light mode the page flickers. Maybe it is possible to implement a server-side saving of the darkmode settings, so that this can be syncedd across devices for a user.

**Comment by thilobillerbeck** [comment]
> One thing about the switching mechanism: I really like what getbootstrap.com is doing in their dropdown. They also include the possibility of "auto" so that the system settings can be (automatically) used. I really love that feature and I think many people will prefer a webpage/app to automatically respect their system preference. image

Sounds good, I will try to port that over. But I think it should stay next to the language selection, since both are related to how the user interface behaves, while the top bar in traewelling is more related to the user context.

**Comment by thilobillerbeck** [comment]
> First of all: Thank you for your work! I love to see, that Trwl is getting an dark mode because of you. <3
> 
> > How do you like the color scheme? What would you see as an improvement?
> 
> In general, I am very satisfied with the DarkMode. I find the readability of the red (trwl red) text on the dark background a bit difficult, but I would put that aside at this point unless you have an idea. 

What do you mean by difficult to read? Is it too popping or do you find it hard to distinguish from the rest of the elements?

> I also found some small things that are not perfect to read, like some buttons.
> 
> image

Oh seems like I have overlooked that.

> Also I am not fully happy with the color difference between the top nav bar, the jumbotron (e.g. on the profile page or the help page) and the dark background. But that are also some things to optimize maybe later step by step?

The problem with the jumbotron is, that the background values are embedded in the HTML while I try to do everything via CSS classes. I checked and don't see any point why it could not be ported over to the SCSS files. Then it would be easier for me to override the styles and it gets a lot easier to maintain anyway.

> Some messages (e.g. the "Checked in" message) are also still a bit light.
> image

Yeah, overlooked that too.

> I would prefer a dropdown like @HerrLevin suggested. Maybe this would fit into the users dropdown in the top navbar?

As said in the comment before, I don't see how this fits logic wise, since all controls regarding the UI are currently situated at the end of the page. Otherwise this gets a full ACK.

> I also noticed that when reloading a page or clicking on it, the dark mode is active by default (css class "dark" is set by default in the HTML). But if you use the light mode the page flickers. Maybe it is possible to implement a server-side saving of the darkmode settings, so that this can be syncedd across devices for a user.

Yeah, that is one of the quirks im currently trying to solve.

**Comment by vainamov** [comment]
grafik
I can not name what exactly bothers me about this, but either the color inside or the border color of the station indicator does not match

**Comment by thilobillerbeck** [comment]
> grafik
> I can not name what exactly bothers me about this, but either the color inside or the border color of the station indicator does not match 


Yeah, that bothers me as well. If you have any idea how to make that look better just let me know. I maybe fiddle around a but with some combinations.

**Comment by thilobillerbeck** [comment]
I know this just kind of extends the described problem, but here we go:

Since this drives me crazy I'll start with the latter problem. The thing with the bulletpoints is, that the red, like it is already noticed,  can look quite agressive in dark scenarios. While the upper variant looks a bit out of place, coloring the whole point red is quite distracting from the rest of the card and draws to much attention for a decorative element imo.
image
The text besides looks like it is some kind of thing besides while it should draw the attention of the reader the most, since it is one of the most important parts of information on this card.

Coloring them in the base color makes them much more muted and the station names can get more attention by the reader. 
image

I will continue to experiment with that. This just kind of describes my problem.

**Comment by thilobillerbeck** [comment]
And to clarify that, the red of the shapes and the progress bar are the same as well as it is with the gray in the connecting line and the circles. The red used for the shaps is darkened already in comparison to the text, since it would appear much more bright otherwise since it covers a lot more space than text and to our eyes appears as a brighter color. That the gray in the circles appears not the same to our eyes as the one in the connecting line is an optical illusion as well.

@vainamov guess that bothers you less?
image

**Comment by vainamov** [comment]
@thilobillerbeck Maybe some part of me doesn't like the proportions of the element in general and it's just now that I realize that 😅

However, could you try `#3a3f4e` for the line and circle background? That's the color you've used for the main card borders in this screen: https://media.discordapp.net/attachments/1117889706097639524/1120133562344161290/localhost_8000_leaderboard.png?width=1872&height=1365

**Comment by thilobillerbeck** [comment]
image

@vainamov this is how it looks with your suggestion

**Comment by thilobillerbeck** [comment]
@MrKrisKrisu, @HerrLevin I applied your suggestions as well, would appreciate a re-review. :)

File: resources/lang/en.json [hunk 1]
```diff
@@ -389,6 +389,10 @@
     "settings.request.accept-success": "Successfully accepted follow request.",
     "settings.request.reject-success": "Successfully rejected follow request.",
     "settings.language.set": "Change language",
+    "settings.colorscheme.set": "Change colorscheme",
+    "settings.colorscheme.light": "Light",
+    "settings.colorscheme.dark": "Dark",
+    "settings.colorscheme.auto": "Sytem preference",
```
Comment (MrKrisKrisu) [inline resources/lang/en.json]:
```suggestion
    "settings.colorscheme.auto": "System preference",
```

**Review by MrKrisKrisu — CHANGES_REQUESTED** [review]

**Comment by MrKrisKrisu** [comment]
> @MrKrisKrisu, @HerrLevin I applied your suggestions as well, would appreciate a re-review. :)

I like the new dropdown. :)
I'm not entirely happy with the position and size - especially in mobile view, but I don't have a good suggestion for this, so... lgtm. <3

**Comment by thilobillerbeck** [comment]
> > @MrKrisKrisu, @HerrLevin I applied your suggestions as well, would appreciate a re-review. :)
> 
> I like the new dropdown. :)
> I'm not entirely happy with the position and size - especially in mobile view, but I don't have a good suggestion for this, so... lgtm. <3

I have something in mind but it is quite hard to implement in Bootstrap since it itself afaik does not ship with media query predicates like Tailwind. My idea would be to switch from row to a column layout on mobile devices and give each dropdown some space.

**Review by MrKrisKrisu — APPROVED** [review]

**Pull request merged**